### ISSUE TITLE
Why disable gradients of on lookup images?

### ISSUE THREAD

**Issue by WBS-123** [nianticlabs/manydepth#17]
In resnet_encoder.py line 275~291
```python
# feature extraction on lookup images - disable gradients to save memory       
with torch.no_grad():            
      if self.adaptive_bins:                
          self.compute_depth_bins(min_depth_bin, max_depth_bin)
     ......
```
I don't understand why disable gradients of on lookup images, if don't do like this, will the result be impacted?

**Comment by mdfirman** [comment]
Thanks for your interest. 

The benefit of doing it is as written in the comment there – we disable the gradients to save memory.

In practice you can enable gradient if you like. We found that it made little difference to scores.

**Comment by WBS-123** [comment]
Thanks for reply！I'll have a try

**Comment by WBS-123** [comment]
> Thanks for your interest.
> 
> The benefit of doing it is as written in the comment there – we disable the gradients to save memory.
> 
> In practice you can enable gradient if you like. We found that it made little difference to scores.

I try to enable gradients of on lookup images, and modifiy the code like this:
in line 276, remove
```python
with torch.no_grad():
```
in line 290~291, remove ```detach()```, like this:
```python
confidence_mask = self.compute_confidence_mask(cost_volume *(1 - missing_mask))
```
is this correct?
and if I enable gradients of on lookup images, the result will get worse?  Should I change the the hyper-parameters such as num_epoch/lr ?

**Comment by mdfirman** [comment]
Yes it looks like those changes should be enough to get gradients to flow to the lookup images.

But I don't know if the result would get better or worse if you do this. You will have to try it out and see!

**Comment by WBS-123** [comment]
> Yes it looks like those changes should be enough to get gradients to flow to the lookup images.
> 
> But I don't know if the result would get better or worse if you do this. You will have to try it out and see!

Thx! If I have any result, I will post here

**Comment by WBS-123** [comment]
After modified the code as above, I got this result:
![2021-06-17 15-16-34 的屏幕截图](https://user-images.githubusercontent.com/49219230/122349696-11b80180-cf7f-11eb-9cfb-37f40ec33cec.png)
it seems like a little worse than origin result.
I'm not sure if it's due to randomness.
Overall, I think the two methods made little difference to score.

**Comment by mdfirman** [comment]
Thanks a lot for reporting back. Glad you could confirm our suspicions.
